### DISCUSSION TITLE
firebase authentication

### DISCUSSION THREAD

**Discussion by harry9216** [kreait/firebase-php#586] [Q&A]
authentication firebase with php
 createuserwithemailandpassword  method is not working , it shows undefined method createuserwithemailandpassword

**Comment by jeromegamez** [comment]
It's hard to help you if you don't share which version of the SDK you're using and don't share the code you tried which is not working, and what is not working 🙏

**Reply by harry9216** [reply]
//this is my controller
public function user_Register(){
				$firebase = $this->firebase->init();
				$userauth = $this->firebase->userauth();
				$this->form_validation->set_rules('fname', 'Fname', 'required');
			$this->form_validation->set_rules('email', 'Email', 'required|valid_email');
			$this->form_validation->set_rules('number', 'Number', 'required|min_length[10]|max_length[12]');
			$this->form_validation->set_rules('address', 'Address', 'required');
			$this->form_validation->set_rules('password', 'Password', 'required|min_length[6]');
			$this->form_validation->set_rules('re-password', 'Re-Password', 'required|matches[password]');				
			$this->form_validation->set_rules('usertype', 'Usertype', 'required');				
				  $config['upload_path'] = './uploads/image';
               	  $config['allowed_types'] =  'jpeg|jpg|png|gif';
                  $config['max_size']  = 1000;
                  $config['file_name']= $this->input->post('images');
                   $this->load->library('upload', $config);
				if (!$this->form_validation->run() || !$this->upload->do_upload('images')) 
				{
					 $error = array('error' => $this->upload->display_errors());
						// $this->load->view('Admin/add_user', $error);
						redirect('add_user',$error);					
				}else{
						$fname = $this->input->post('fname');
						$email = $this->input->post('email');
						$number = $this->input->post('number');					
						$address = $this->input->post('address');
						$password = $this->input->post('password');
						// $repassword = $this->input->post('re-password');
						$type = $this->input->post('usertype');
						$images= $this->upload->data();
						$img = $images['file_name'];
						 $user = $userauth->createUserWithEmailAndPassword($email,$password);
						
						$data = [
							'email' => $email,
							'fullname' => $fname,
							'number' => $number,
							'address' => $address,
							'password' => md5($repassword),
							'profile' => $img,
							'type'    => $type,
							'status' => 1,
							'created_at' => date('Y-m-d H:i:s'),
							'updated_at'=>date('Y-m-d H:i:s')						
						];
					$check = $this->Adminmodel->user_register($data);
					if ($check == true) {
							echo " <script> alert('Something is wrong') </script>" ; 
							
					}else{				
				    echo "<script> alert('successfully added ') </script>"; 
					redirect('all_user');	
					}
			}
}

**Reply by harry9216** [reply]
<?php
defined('BASEPATH') OR exit('No direct script access allowed');

require __DIR__.'/vendor/autoload.php';
use Kreait\Firebase\Factory;
use Kreait\Firebase\ServiceAccount;
use Kreait\Firebase\Auth;
use Google\Cloud\Storage\StorageClient;

class Firebase{
    protected $config   = array();
    protected $serviceAccount;
    //  private $_baseURI;
    // private $_timeout = 10;
    // private $_token;
    public function __construct()
    {
        // Assign the CodeIgniter super-object
        $this->CI =& get_instance();


      //$serviceAccount  = ServiceAccount::fromJsonFile(__DIR__.'/woofie-84481-firebase-adminsdk-tc2d4-7765067003.json');        
    }
    public function init()
    {
       return  $firebase = (new Factory)
        ->withServiceAccount(__DIR__.'/woofie-84481-firebase-adminsdk-tc2d4-7765067003.json')
        ->withDatabaseUri('https://woofie-84481-default-rtdb.firebaseio.com')
        ->createDatabase();
    }
    public function userauth()
    {
       return  $firebase = (new Factory)
        ->withServiceAccount(__DIR__.'/woofie-84481-firebase-adminsdk-tc2d4-7765067003.json')
        ->withDatabaseUri('https://woofie-84481-default-rtdb.firebaseio.com')
        ->createDatabase();

    }



   
}

?>

**Reply by harry9216** [reply]
im using Version 5.14

**Reply by harry9216** [reply]
sir is this code understandable or not?

**Reply by jeromegamez** [reply]
I'm not a Sir 😅, and it would be nice if you could use Markdown code fences when pasting code 🙏 

```php
$firebase = $this->firebase->init();
$userauth = $this->firebase->userauth();
...
$user = $userauth->createUserWithEmailAndPassword($email,$password);
```

I don't know what `$this->firebase` is or what `$this->firebase->userauth()` returns, but that's where you should look. The method does exist in the `Kreait\Firebase\Auth` class: https://github.com/kreait/firebase-php/blob/8b45c07d922364ba9298fa07cbe7ea676c1d05f4/src/Firebase/Auth.php#L163

**Reply by harry9216** [reply]
This method  exist in the Kreait\Firebase\Auth class:

 public function createUserWithEmailAndPassword($email, $password): UserRecord
    {
        return $this->createUser(Request\CreateUser::new()
            ->withUnverifiedEmail($email)
            ->withClearTextPassword($password)
        );
    }

**Reply by harry9216** [reply]
should i need this code ( $auth = $factory->createAuth(); )   to define in my code  before using  this code
 $user = $userauth->createUserWithEmailAndPassword($email,$password);

**Reply by jeromegamez** [reply]
I'm not sure what you need, it depends on what is in `$userauth`. You could try doing a `var_dump(get_class($userauth)); exit;` and look in the corresponding class/file what is done there. But yes, you will need `$factory->createAuth();` to be able to use the `createUser...` method 🤞

**Reply by harry9216** [reply]
thanks a lot dear

**Reply by harry9216** [reply]
Message: Call to undefined method Kreait\Firebase\Database::createAuth()

**Reply by harry9216** [reply]
Message: Call to undefined method Kreait\Firebase\Database::createUserWithEmailAndPassword()

**Reply by jeromegamez** [reply]
Message: Call to undefined method Jerome::guessReasonForErrorWithoutContext()

😅😂

**Reply by jeromegamez** [reply]
You're trying to call a method on the Database class - those methods only exist on the Auth class

**Reply by harry9216** [reply]
why is this error 

Message: Call to undefined method Kreait\Firebase\Database::createAuth()

its such a headache. .  help me out

**Reply by jeromegamez** [reply]
The error is quite clear already - you're trying to call `createAuth()` on an instance of `Kreait\Firebase\Database`, but this method does not exist. `createAuth()` only exists on the `Kreait\Firebase\Factory` class.

I'm getting the feeling that you followed an (outdated) tutorial on the internet/youtube, and that's where the mix-ups come from.

**Reply by harry9216** [reply]
what should i do now ? 
any suggestion
